connector/google/google.go

```diff
@@ -0,0 +1,194 @@
+)
+// Config holds configuration options for OpenID Connect logins.
+type Config struct {
+	Issuer       string `json:"issuer"`
+	ClientID     string `json:"clientID"`
+	ClientSecret string `json:"clientSecret"`
+	RedirectURI  string `json:"redirectURI"`
+	Scopes []string `json:"scopes"`
+	// Optional list of whitelisted domains when using Google
+	// If this field is nonempty, only users from a listed domain will be allowed to log in
+	HostedDomains []string `json:"hostedDomains"`
+	filePath  string `json:"serviceAccountCred"`
+	Subject         string `json:"subject"`
+	GroupPrefix	string `json:"groupPrefix"`
+}
+
+// Open returns a connector which can be used to login users through an upstream
+// OpenID Connect provider.
+func (c *Config) Open(logger logrus.FieldLogger) (conn connector.Connector, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	provider, err := oidc.NewProvider(ctx, c.Issuer)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to get provider: %v", err)
+	}
+	scopes := []string{oidc.ScopeOpenID}
+	if len(c.Scopes) > 0 {
+		scopes = append(scopes, c.Scopes...)
+	} else {
+		scopes = append(scopes, "profile", "email")
+	}
+	//creating client for extraxting groups for users
+	clientID := c.ClientID
+	contents, err := ioutil.ReadFile(c.filePath)
+	b := []byte(contents)
+	config, err := google.JWTConfigFromJSON(b,
+		admin.AdminDirectoryGroupReadonlyScope,
+	)
+	config.Subject = c.Subject
+	if err != nil {
+		fmt.Errorf("Unable to parse client secret file to config: %v", err)
+	}
+	client, err := admin.New(config.Client(context.Background()))
+	return &googleConnector{
+		redirectURI: c.RedirectURI,
+		oauth2Config: &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: c.ClientSecret,
+			Endpoint:     provider.Endpoint(),
+			Scopes:       scopes,
+			RedirectURL:  c.RedirectURI,
+		},
+		verifier: provider.Verifier(
+			&oidc.Config{ClientID: clientID},
+		),
+		logger:        logger,
+		cancel:        cancel,
+		hostedDomains: c.HostedDomains,
+		client: client,
+		groupPrefix:	c.GroupPrefix,
+	}, nil
+}
+var (
+	_ connector.CallbackConnector = (*googleConnector)(nil)
+	_ connector.RefreshConnector  = (*googleConnector)(nil)
+)
+type googleConnector struct {
+	redirectURI   string
+	oauth2Config  *oauth2.Config
+	verifier      *oidc.IDTokenVerifier
+	ctx           context.Context
+	cancel        context.CancelFunc
+	logger        logrus.FieldLogger
+	hostedDomains []string
+	client		*admin.Service
+	groupPrefix	string
+}
+func (c *googleConnector) Close() error {
+	c.cancel()
+	return nil
+}
+func (c *googleConnector) LoginURL(s connector.Scopes, callbackURL, state string) (string, error) {
+	if c.redirectURI != callbackURL {
+		return "", fmt.Errorf("expected callback URL %q did not match the URL in the config %q", callbackURL, c.redirectURI)
+	}
+	if len(c.hostedDomains) > 0 {
+		preferredDomain := c.hostedDomains[0]
+		if len(c.hostedDomains) > 1 {
+			preferredDomain = "*"
+		}
```

Comment:
This stuff can also get reduced/cleaned up if `hostedDomain` is just a single string. preferredDomain can go away and just be replaced with `hostedDomain`